extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java

```diff
@@ -532,4 +548,113 @@ private String serializerDeserializerFor(DefaultSerdeDiscoveryState discovery, T
+    }
+
+    // visible for testing
+    void processIncomingChannelForReflectiveClassPayload(IndexView index, Config config,
+            BiConsumer<AnnotationInstance, Type> annotationAcceptor) {
+        processAnnotationsForReflectiveClassPayload(index, config, DotNames.CHANNEL, false,
+                annotation -> getIncomingChannelType(getInjectionPointTypeFromChannel(annotation)),
+                annotationAcceptor);
+    }
+
+    private void processAnnotationsForReflectiveClassPayload(IndexView index, Config config, DotName annotationType,
+            boolean serializer, Function<AnnotationInstance, Type> typeExtractor,
+            BiConsumer<AnnotationInstance, Type> annotationAcceptor) {
+        for (AnnotationInstance annotation : index.getAnnotations(annotationType)) {
+            String channelName = annotation.value().asString();
+            Type type = typeExtractor.apply(annotation);
+            extractKeyValueType(type, (key, value) -> {
+                if (key != null && isSerdeJson(index, config, channelName, serializer, true)) {
+                    annotationAcceptor.accept(annotation, key);
+                }
+                if (value != null && isSerdeJson(index, config, channelName, serializer, false)) {
+                    annotationAcceptor.accept(annotation, value);
+                }
+            });
+        }
+    }
+
+    private boolean isSerdeJson(IndexView index, Config config, String channelName, boolean serializer, boolean isKey) {
+        ConfigValue configValue = config.getConfigValue(getConfigName(channelName, serializer, isKey));
+        if (configValue.getValue() != null) {
+            DotName serdeName = DotName.createSimple(configValue.getValue());
+            return serializer ? isSubclassOfJsonSerializer(index, serdeName) : isSubclassOfJsonDeserializer(index, serdeName);
+        }
+        return false;
+    }
+
+    String getConfigName(String channelName, boolean serializer, boolean isKey) {
+        return "mp.messaging." +
+                (serializer ? "outgoing" : "incoming") + "." +
+                channelName + "." +
+                (isKey ? "key" : "value") + "." +
+                (serializer ? "serializer" : "deserializer");
+    }
+
+    private boolean isSubclassOfJsonSerializer(IndexView index, DotName serializerName) {
+        return getSubclassOfWithTypeArgument(index, DotNames.OBJECT_MAPPER_SERIALIZER, serializerName) ||
+                getSubclassOfWithTypeArgument(index, DotNames.JSONB_SERIALIZER, serializerName);
+    }
+
+    private boolean isSubclassOfJsonDeserializer(IndexView index, DotName serializerName) {
+        return getSubclassOfWithTypeArgument(index, DotNames.OBJECT_MAPPER_DESERIALIZER, serializerName) ||
+                getSubclassOfWithTypeArgument(index, DotNames.JSONB_DESERIALIZER, serializerName);
+    }
+
+    private boolean getSubclassOfWithTypeArgument(IndexView index, DotName superclass, DotName expectedType) {
```

Comment:
The name here is copied from a method that does something different I think. Rename to `isSubclassOf`?

Reply:
I agree!